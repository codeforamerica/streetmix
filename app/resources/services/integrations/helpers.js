const { User } = require('../../../db/models')

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
const findUser = async function (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
const addOrUpdateByProviderName = function (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
const syncAccountStatus = async function (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

module.exports = { findUser, addOrUpdateByProviderName, syncAccountStatus }
