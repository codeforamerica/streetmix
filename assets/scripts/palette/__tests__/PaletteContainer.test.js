/* eslint-env jest */
import React from 'react'
import { screen } from '@testing-library/react'
import { renderWithReduxAndIntl } from '../../../../test/helpers/render'
import PaletteContainer from '../PaletteContainer'

jest.mock('../../segments/view', () => {
  const actual = jest.requireActual('../../segments/view')
  return {
    ...actual,
    drawSegmentContents: jest.fn()
  }
})

describe('PaletteContainer', () => {
  it('renders', () => {
    const { asFragment } = renderWithReduxAndIntl(<PaletteContainer />, {
      initialState: {
        app: {
          everythingLoaded: true
        },
        // Assume the environment editor is enabled. This will render
        // the <EnvironmentEditor /> button for test coverage.
        flags: {
          ENVIRONMENT_EDITOR: { value: true },
          DEBUG_SEGMENT_CANVAS_RECTANGLES: { value: false } // for SegmentForPalette
        }
      }
    })

    expect(asFragment()).toMatchSnapshot()
  })

  it('doesn’t render in read-only mode', () => {
    const { container } = renderWithReduxAndIntl(<PaletteContainer />, {
      initialState: {
        app: {
          readOnly: true
        }
      }
    })

    expect(container.firstChild).toBe(null)
  })

  it("doesn’t render items when sprites haven't loaded", () => {
    renderWithReduxAndIntl(<PaletteContainer />, {
      initialState: {
        app: {
          everythingLoaded: false
        }
      }
    })

    expect(screen.queryByRole('list')).toBe(null)
  })
})
