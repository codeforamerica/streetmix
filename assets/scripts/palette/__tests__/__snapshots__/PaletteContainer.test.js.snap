// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteContainer renders 1`] = `
<DocumentFragment>
  <div
    class="palette-container"
  >
    <div
      class="palette-commands palette-commands-left"
    >
      <button
        title="Toggle tools"
      >
        <svg
          class="svg-inline--fa fa-tools"
        />
      </button>
      <div
        class="supermoon-tooltip"
      >
        <button
          class="close"
          title="Dismiss"
        >
          <svg
            class="svg-inline--fa fa-times"
          />
        </button>
        <p>
          <strong>
            You’ve got some new tools!‎
          </strong>
           Click on this button to activate some new abilities.‎
        </p>
        <div
          class="palette-tooltip-pointer-container"
        >
          <div
            class="palette-tooltip-pointer"
          />
        </div>
      </div>
    </div>
    <div
      class="palette-items-scrollable-container"
    >
      <div
        class="palette-items"
      >
        <ul>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 48px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 48px; height: 445px; left: 0px;"
                width="48"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 40px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 40px; height: 445px; left: 0px;"
                width="40"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 44px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 44px; height: 445px; left: 0px;"
                width="44"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 40px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 40px; height: 445px; left: 0px;"
                width="40"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette segment-disabled"
            data-testid="segment-for-palette"
            style="width: 56px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 56px; height: 445px; left: 0px;"
                width="56"
              />
            </div>
            <svg
              class="svg-inline--fa fa-lock"
            />
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 40px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 40px; height: 445px; left: 0px;"
                width="40"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 40px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 40px; height: 445px; left: 0px;"
                width="40"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 56px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 56px; height: 445px; left: 0px;"
                width="56"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 32px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 32px; height: 445px; left: 0px;"
                width="32"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette segment-disabled"
            data-testid="segment-for-palette"
            style="width: 32px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 32px; height: 445px; left: 0px;"
                width="32"
              />
            </div>
            <svg
              class="svg-inline--fa fa-lock"
            />
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 48px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 48px; height: 445px; left: 0px;"
                width="48"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 52px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 52px; height: 445px; left: 0px;"
                width="52"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 64px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 64px; height: 445px; left: 0px;"
                width="64"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 52px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 52px; height: 445px; left: 0px;"
                width="52"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 40px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 40px; height: 445px; left: 0px;"
                width="40"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 52px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 52px; height: 445px; left: 0px;"
                width="52"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 64px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 64px; height: 445px; left: 0px;"
                width="64"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 64px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 64px; height: 445px; left: 0px;"
                width="64"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 72px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 72px; height: 445px; left: 0px;"
                width="72"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 72px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 72px; height: 445px; left: 0px;"
                width="72"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 72px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 72px; height: 445px; left: 0px;"
                width="72"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette"
            data-testid="segment-for-palette"
            style="width: 60px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 60px; height: 445px; left: 0px;"
                width="60"
              />
            </div>
          </li>
          <li
            class="segment segment-in-palette segment-disabled"
            data-testid="segment-for-palette"
            style="width: 64px;"
          >
            <div
              style="height: 80px;"
            >
              <canvas
                class="segment-image"
                height="445"
                style="width: 64px; height: 445px; left: 0px;"
                width="64"
              />
            </div>
            <svg
              class="svg-inline--fa fa-lock"
            />
          </li>
        </ul>
      </div>
      <button
        class="scrollable scroll-left"
        style="left: -15px; opacity: 0; pointer-events: none;"
      >
        <svg
          class="svg-inline--fa fa-chevron-left"
        />
      </button>
      <button
        class="scrollable scroll-right"
        style="right: -15px; opacity: 0; pointer-events: none;"
      >
        <svg
          class="svg-inline--fa fa-chevron-right"
        />
      </button>
    </div>
    <div
      class="palette-commands palette-commands-right"
    >
      <button
        disabled=""
        title="Undo"
      >
        <svg
          class="svg-inline--fa fa-undo"
        />
      </button>
      <button
        disabled=""
        title="Redo"
      >
        <svg
          class="svg-inline--fa fa-redo"
        />
      </button>
    </div>
    <div
      class="palette-trashcan"
      hidden=""
    >
      Drag here to remove
    </div>
  </div>
</DocumentFragment>
`;
